service/src/identity/http/accounts.rs

```diff
@@ -0,0 +1,245 @@
+
+    Ok(Json(SignupResponse {
+        account_id,
+        device_id: prepared.device_id,
+        root_kid: prepared.root_kid,
+    }))
+}
+
+pub(crate) fn extract_device_id(envelope: &SignedEnvelope) -> Result<Uuid, (StatusCode, String)> {
+    let value = envelope.payload.get("device_id").ok_or_else(|| {
+        (
+            StatusCode::BAD_REQUEST,
+            "device_id missing in payload".to_string(),
+        )
+    })?;
+    match value {
+        serde_json::Value::String(s) => Uuid::parse_str(s).map_err(|_| {
+            (
+                StatusCode::BAD_REQUEST,
+                "device_id must be a UUID".to_string(),
+            )
+        }),
+        _ => Err((
+            StatusCode::BAD_REQUEST,
+            "device_id must be a string".to_string(),
+        )),
+    }
+}
+
+pub(crate) fn internal_error<E: std::fmt::Display>(err: E) -> (StatusCode, String) {
+    (StatusCode::INTERNAL_SERVER_ERROR, err.to_string())
+}
+
+pub(crate) fn decode_key(encoded: &str, field: &str) -> Result<Vec<u8>, (StatusCode, String)> {
+    base64::engine::general_purpose::URL_SAFE_NO_PAD
+        .decode(encoded.as_bytes())
+        .map_err(|_| (StatusCode::BAD_REQUEST, format!("invalid {field} encoding")))
+}
+
+fn prepare_signup_request(payload: SignupRequest) -> Result<PreparedSignup, (StatusCode, String)> {
+    let username = payload.username.trim().to_lowercase();
+    if username.is_empty() {
+        return Err((StatusCode::BAD_REQUEST, "username is required".to_string()));
+    }
+
+    let root_pubkey_bytes = decode_key(&payload.root_pubkey, "root_pubkey")?;
+    let root_kid = derive_kid(&root_pubkey_bytes);
+
+    // Basic envelope checks
+    let device_id = extract_device_id(&payload.delegation_envelope)?;
+    let account_id = payload
+        .delegation_envelope
+        .signer
+        .account_id
+        .ok_or_else(|| {
+            (
+                StatusCode::BAD_REQUEST,
+                "delegation signer must include account_id".to_string(),
+            )
+        })?;
+    if payload.delegation_envelope.signer.kid != root_kid {
+        return Err((
+            StatusCode::BAD_REQUEST,
+            "delegation signer kid does not match root pubkey".to_string(),
+        ));
+    }
+
+    verify_envelope(&payload.delegation_envelope, &root_pubkey_bytes).map_err(|err| {
+        (
+            StatusCode::BAD_REQUEST,
+            format!("invalid delegation signature: {err}"),
+        )
+    })?;
+
+    let device_pubkey_bytes = decode_key(&payload.device_pubkey, "device_pubkey")?;
+    let device_kid = derive_kid(&device_pubkey_bytes);
+
+    let (device_name, device_type) = payload.device_metadata.map_or_else(
+        || (None, "other".to_string()),
+        |meta| {
+            let name = meta.name;
+            let device_type = meta.r#type.unwrap_or_else(|| "other".to_string());
+            (name, device_type)
+        },
+    );
+
+    Ok(PreparedSignup {
+        username,
```

Comment:
this is hard to read without attribute names. I think we should add a linting rule to ensure these always have attribute names.